content/oci/deletableoci.go

```diff
@@ -0,0 +1,216 @@
+			ds.tagResolver.Untag(reference)
+			untagged = true
+		}
+	}
+	if err := ds.graph.Remove(ctx, target); err != nil {
+		return err
+	}
+	if untagged && ds.AutoSaveIndex {
+		err := ds.saveIndex()
+		if err != nil {
+			return err
+		}
+	}
+
+	// we may not need the type assertion, if we change Store.storage from the
+	// read-only interface to a struct pointer.
+	storage := ds.storage.(*Storage)
+	return storage.Delete(ctx, target)
+}
+
+// Exists returns true if the described content exists.
+func (ds *DeletableStore) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Exists(ctx, target)
+}
+
+// Tag tags a descriptor with a reference string.
+// reference should be a valid tag (e.g. "latest").
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md#indexjson-file
+func (ds *DeletableStore) Tag(ctx context.Context, desc ocispec.Descriptor, reference string) error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	return ds.Store.Tag(ctx, desc, reference)
+}
+
+// Resolve resolves a reference to a descriptor. If the reference to be resolved
+// is a tag, the returned descriptor will be a full descriptor declared by
+// github.com/opencontainers/image-spec/specs-go/v1. If the reference is a
+// digest the returned descriptor will be a plain descriptor (containing only
+// the digest, media type and size).
+func (ds *DeletableStore) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Resolve(ctx, reference)
+}
+
+// Predecessors returns the nodes directly pointing to the current node.
+// Predecessors returns nil without error if the node does not exists in the
+// store.
+func (ds *DeletableStore) Predecessors(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Predecessors(ctx, node)
+}
+
+// Tags lists the tags presented in the `index.json` file of the OCI layout,
+// returned in ascending order.
+// If `last` is NOT empty, the entries in the response start after the tag
+// specified by `last`. Otherwise, the response starts from the top of the tags
+// list.
+//
+// See also `Tags()` in the package `registry`.
+func (ds *DeletableStore) Tags(ctx context.Context, last string, fn func(tags []string) error) error {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Tags(ctx, last, fn)
+}
+
+// SaveIndex writes the `index.json` file to the file system.
+//   - If AutoSaveIndex is set to true (default value),
+//     the OCI store will automatically save the index on each Tag() call.
+//   - If AutoSaveIndex is set to false, it's the caller's responsibility
+//     to manually call this method when needed.
+func (ds *DeletableStore) SaveIndex() error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	return ds.saveIndex()
+}
```

Comment:
not sure about this function. Should I use `Store.SaveIndex()`?

Reply:
Why not :D

They look exactly the same.

Reply:
We can refactor `Store.SaveIndex()` and make it
```go
func (s *Store) SaveIndex() error {
	s.indexLock.Lock()
	defer s.indexLock.Unlock()

	return s.saveIndex()
}

So that we can reuse the `saveIndex` here.
```

Reply:
> Why not :D
> 
> They look exactly the same.

`Store.SaveIndex()` has another lock, I worry about lock complexities.

Reply:
> We can refactor `Store.SaveIndex()` and make it
> 
> ```go
> func (s *Store) SaveIndex() error {
> 	s.indexLock.Lock()
> 	defer s.indexLock.Unlock()
> 
> 	return s.saveIndex()
> }
> ```
> 
> So that we can reuse the `saveIndex()` here.

Done it this way.

Reply:
> We can refactor `Store.SaveIndex()` and make it
> 
> ```go
> func (s *Store) SaveIndex() error {
> 	s.indexLock.Lock()
> 	defer s.indexLock.Unlock()
> 
> 	return s.saveIndex()
> }
> ```
> 
> So that we can reuse the `saveIndex()` here.

Done it this way.